include/bredis/impl/connection.ipp

```diff
@@ -60,10 +60,8 @@ Connection<NextLayer>::async_read(DynamicBuffer &rx_buff,
     asio::async_result<real_handler_t> async_result(real_handler);
 
     async_read_until(stream_, rx_buff, MatchResult<Iterator>(replies_count), [
-        handler = std::move(real_handler), &rx_buff, replies_count
-    ](const sys::error_code &error_code, std::size_t bytes_transferred) {
-
-        real_handler_t real_handler(std::move(handler));
+        real_handler = std::move(real_handler), &rx_buff, replies_count
+    ](const sys::error_code &error_code, std::size_t bytes_transferred) mutable {
```

Comment:
If `real_handler` has overloaded `asio_handler_invoke` to dispatch using an `io_service::strand`, this code will cause the `stream_` to be accessed in an unsafe way if there is more than one thread calling `io_service::run`. That is because the lambda you provide, does not have the necessary overload of `asio_handler_invoke`.

From
http://www.boost.org/doc/libs/1_59_0/doc/html/boost_asio/reference/asio_handler_invoke.html

_"When asynchronous operations are composed from other asynchronous operations, all intermediate handlers should be invoked using the same method as the final handler. This is required to ensure that user-defined objects are not accessed in a way that may violate the guarantees."_